crates/ibc/src/core/ics02_client/events.rs

```diff
@@ -21,262 +21,324 @@ pub const CLIENT_TYPE_ATTRIBUTE_KEY: &str = "client_type";
-///
-/// # Note
-/// The parsing of `Key`s and `Value`s never fails, because the
-/// `FromStr` instance of `tendermint::abci::tag::{Key, Value}`
-/// is infallible, even if it is not represented in the error type.
-/// Once tendermint-rs improves the API of the `Key` and `Value` types,
-/// we will be able to remove the `.parse().unwrap()` calls.
-impl From<Attributes> for Vec<Tag> {
-    fn from(attrs: Attributes) -> Self {
-        let client_id = Tag {
-            key: CLIENT_ID_ATTRIBUTE_KEY.parse().unwrap(),
-            value: attrs.client_id.to_string().parse().unwrap(),
-        };
-        let client_type = Tag {
-            key: CLIENT_TYPE_ATTRIBUTE_KEY.parse().unwrap(),
-            value: attrs.client_type.as_str().parse().unwrap(),
-        };
-        let consensus_height = Tag {
-            key: CONSENSUS_HEIGHT_ATTRIBUTE_KEY.parse().unwrap(),
-            value: attrs.consensus_height.to_string().parse().unwrap(),
-        };
-        vec![client_id, client_type, consensus_height]
+#[derive(Debug)]
+struct HeaderAttribute {
+    header: Any,
+}
+
+impl From<HeaderAttribute> for Tag {
+    fn from(attr: HeaderAttribute) -> Self {
+        Tag {
+            key: HEADER_ATTRIBUTE_KEY.parse().unwrap(),
+            value: String::from_utf8(hex::encode(attr.header.value))
+                .unwrap()
+                .parse()
+                .unwrap(),
+        }
+    }
+}
+
+impl From<Any> for HeaderAttribute {
+    fn from(header: Any) -> Self {
+        Self { header }
     }
 }
 
 /// CreateClient event signals the creation of a new on-chain client (IBC client).
-#[derive(Debug, Serialize, Clone, PartialEq, Eq)]
-pub struct CreateClient(pub Attributes);
+#[derive(Debug)]
+pub struct CreateClient {
+    client_id: ClientIdAttribute,
+    client_type: ClientTypeAttribute,
+    consensus_height: ConsensusHeightAttribute,
+}
 
 impl CreateClient {
+    pub fn new(client_id: ClientId, client_type: ClientType, consensus_height: Height) -> Self {
+        Self {
+            client_id: ClientIdAttribute::from(client_id),
+            client_type: ClientTypeAttribute::from(client_type),
+            consensus_height: ConsensusHeightAttribute::from(consensus_height),
+        }
+    }
+
     pub fn client_id(&self) -> &ClientId {
-        &self.0.client_id
+        &self.client_id.client_id
     }
-}
 
-impl Display for CreateClient {
-    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FmtError> {
-        write!(f, "CreateClient {{ {} }}", self.0)
+    pub fn client_type(&self) -> &ClientType {
+        &self.client_type.client_type
     }
-}
 
-impl From<Attributes> for CreateClient {
-    fn from(attrs: Attributes) -> Self {
-        CreateClient(attrs)
+    pub fn consensus_height(&self) -> &Height {
+        &self.consensus_height.consensus_height
     }
 }
 
-impl From<CreateClient> for IbcEvent {
-    fn from(v: CreateClient) -> Self {
-        IbcEvent::CreateClient(v)
+impl Display for CreateClient {
```

Comment:
Noticed that the `Display` implementation has changed. Not sure if we still need a `Display` implementation - might we be able to remove the `IbcEvent`'s `Display` implementation altogether?

Reply:
Agreed [31301ab](https://github.com/cosmos/ibc-rs/pull/138/commits/31301ab18488130856c8fe920a262df0023dd684)